File: tests/test_custom_base.cpp
/*
    tests/test_custom_base.cpp -- test custom type hierarchy support

    All rights reserved. Use of this source code is governed by a
    BSD-style license that can be found in the LICENSE file.
*/

#include "pybind11_tests.h"

namespace {

struct Base {
    int i = 5;
};

struct Derived {
    int j = 6;

    // just to prove the base can be anywhere
    Base base;
};

} // namespace

TEST_SUBMODULE(custom_base, m) {

    py::class_<Base>(m, "Base").def_readwrite("i", &Base::i);

    py::class_<Derived>(m, "Derived", py::custom_base<Base>([](void *o) -> void * {
                            return &reinterpret_cast<Derived *>(o)->base;
                        }))
        .def_readwrite("j", &Derived::j);

    m.def("create_derived", []() { return new Derived; });
    m.def("create_base", []() { return new Base; });

    m.def("base_i", [](Base *b) { return b->i; });

    m.def("derived_j", [](Derived *d) { return d->j; });
};

File: tests/test_custom_base.py
from pybind11_tests import custom_base as m


def test_cb_base():
    b = m.create_base()

    assert isinstance(b, m.Base)
    assert b.i == 5

    assert m.base_i(b) == 5


def test_cb_derived():
    d = m.create_derived()

    assert isinstance(d, m.Derived)
    assert isinstance(d, m.Base)

    assert d.i == 5
    assert d.j == 6

    assert m.base_i(d) == 5
    assert m.derived_j(d) == 6
